### PR TITLE
[ELF] Postpone "unable to move location counter backward" error

### LINKED ISSUE #66836: [LLD] LLD can report "unable to move location counter backward" error too early
I have the following linker script:

```
MEMORY
{
  ...
  RAM_DATA               (RW)  : ORIGIN = 0x20008000, LENGTH = 0x004F8000
  ...
}

ram_data_end      = ORIGIN(RAM_DATA) + LENGTH(RAM_DATA);

SECTIONS {
  ...  // other output sections like text, rodata, etc
  __etext = .;
  __DATA_ROM = .;
  __end_of_code_in_ram_data =
      ABSOLUTE(__etext - ORIGIN(RAM_CODE) + ORIGIN(RAM_DATA));
  .data __end_of_code_in_ram_data : AT(__DATA_ROM) ALIGN(4) {
    __data_start__ = .;
    *(.data);
    *(.data*);
    ...  // stuff placed here
    __data_end__ = .;        /* define a global symbol at data end */
  } > RAM_DATA

  .shared_ram.unused_space (NOLOAD) : ALIGN(4) {
    . = ABSOLUTE(ram_data_end);
  } >RAM_DATA

  ... // other output sections
}
```

which would often report this error: `unable to move location counter backward for: .shared_ram.unused_space`. This is because on the first call to `script->assignAddresses();` in `finalizeAddressDependentContent`, all of the input data sections placed into `.data` exceeds the remaining space for RAM_DATA, so by the time we reach `. = ABSOLUTE(ram_data_end);`, DOT is well past the end of RAM_DATA.

However, on subsequent calls to `script->assignAddresses()` in the fixed-point loop in `finalizeAddressDependentContent`, the start of `.data` is changed and a smaller value than the first iteration because something was changed in previous sections. This could be from linker relaxations, sections being removed, allocation sizes changing, etc. Because the starting address of `.data` is smaller but all the input data sections remain the same, it's possible for subsequent runs of `script->assignAddresses()` to not throw the same error because it happens to fit.

The issue here is that lld will still throw this error from the first run even though all the sections can fit as expected in the final memory region after the fixed point loop has finished. I think the right approach here is that LLD should only return this error if DOT ends up moving backwards for the final call to this function, similar to how we delay checking of memory region sizes until the very end of linking.

TODO: Come up with a public-facing minimal reproducer. I haven't made one yet because it's too hard.

### PR DESCRIPTION
**Pull request by MaskRay**
The size of .ARM.exidx may shrink across `assignAddress` calls. It is possible
that the initial iteration has a larger location counter, causing `__code_size =
__code_end - .; osec : { . += __code_size; }` to report an error, while the error would
have been suppressed for subsequent `assignAddress` iterations.

Other sections like .relr.dyn may change sizes across `assignAddress` calls as
well. However, their initial size is zero, so it is difficiult to trigger a
similar error.

Similar to https://reviews.llvm.org/D152170, postpone the error reporting.
Fix #66836. While here, add more information to the error message.

### CODE REVIEW

File: lld/ELF/LinkerScript.cpp [hunk 1]
```diff
@@ -169,9 +169,16 @@ void LinkerScript::expandOutputSection(uint64_t size) {
...
-  if (val < dot && inSec)
-    error(loc + ": unable to move location counter backward for: " +
-          state->outSec->name);
+  // If val is smaller and we are in an output section, record the error and
+  // report it if this is the last assignAddresses iteration. dot may be smaller
+  // if there is another assignAddresses iteration.
+  if (val < dot && inSec) {
+    backwardDotErr =
+        (loc + ": unable to move location counter (0x" + Twine::utohexstr(dot) +
+         ") backward (0x" + Twine::utohexstr(val) + ") for section '" +
```
Comment (arichardson) [inline lld/ELF/LinkerScript.cpp:178]: Much prefer the new error message but I noticed that it's ambiguous if this means we are moving it backwards by a number of bytes or to given address. How about the following:

```suggestion
         ") backward to (0x" + Twine::utohexstr(val) + ") for section '" +
```

File: lld/test/ELF/linkerscript/locationcountererr.test [hunk 2]
```diff
@@ -1,7 +1,7 @@
 # REQUIRES: x86
 # RUN: llvm-mc -filetype=obj -triple=x86_64-unknown-linux /dev/null -o %t
 # RUN: not ld.lld %t --script %s -o /dev/null 2>&1 | FileCheck %s
-# CHECK: {{.*}}.test:8: unable to move location counter backward for: .text
+# CHECK: {{.*}}.test:8: unable to move location counter (0x2000) backward (0x10) for section '.text'
```
Comment (arichardson) [inline lld/test/ELF/linkerscript/locationcountererr.test:4]: Maybe worth adding a test with more than one backwards error? Previously we'd get multiple reports now just the first one. There probably isn't much value in reporting further errors, but I think it does make sense to document this behaviour in a test?

**Review by arichardson — APPROVED** [review]
Looks sensible to me but probably a good idea to get another review.

File: lld/ELF/LinkerScript.cpp (hunk 1 above)
Comment (MaskRay) [reply lld/ELF/LinkerScript.cpp]: Updated in  54daea8e15f7e5e95e6531d34616062af81c4d84 to `backward to 0x... for section 'xxx'`

File: lld/test/ELF/linkerscript/locationcountererr.test (hunk 2 above)
Comment (MaskRay) [reply lld/test/ELF/linkerscript/locationcountererr.test]: Thanks for the suggestion. 54daea8e15f7e5e95e6531d34616062af81c4d84 added two more memory regions to demonstrate the full set of errors.

**Review by smithp35 — APPROVED** [review]
LGTM too.

File: lld/ELF/LinkerScript.h [hunk 3]
```diff
@@ -345,8 +345,8 @@ class LinkerScript final {
   // Describe memory region usage.
   void printMemoryUsage(raw_ostream &os);
 
-  // Verify memory/lma overflows.
-  void checkMemoryRegions() const;
+  // Check backward location counter assignment and memory region/LMA overflows.
+  void checkDotAndMemoryRegions() const;
```
[the remark this answers is not in the document]
Comment (MaskRay) [reply lld/ELF/LinkerScript.h]: Thanks for the suggestion. I remember that I've seen similar `ASSERT` issues. We can postpone the check when the problem arises...

[the remark this answers is not in the document]
Comment (MaskRay) [reply lld/ELF/LinkerScript.h]: Renamed to checkFinalScriptConditions in 2bf25dc88bfa848ca520578ce2c0e6bb15d1a1a9

**Pull request merged**